src/execution/operator/join/physical_iejoin.cpp

```diff
@@ -0,0 +1,1415 @@
+	}
+
+	void GetNextPair(ClientContext &client, IEJoinGlobalState &gstate, IEJoinLocalSourceState &lstate) {
+		auto &left_table = *gstate.tables[0];
+		auto &right_table = *gstate.tables[1];
+
+		const auto left_blocks = left_table.BlockCount();
+		const auto right_blocks = right_table.BlockCount();
+		const auto pair_count = left_blocks * right_blocks;
+
+		// Regular block
+		const auto i = next_pair++;
+		if (i < pair_count) {
+			const auto b1 = i / right_blocks;
+			const auto b2 = i % right_blocks;
+
+			lstate.left_block_index = b1;
+			lstate.left_base = left_bases[b1];
+
+			lstate.right_block_index = b2;
+			lstate.right_base = right_bases[b2];
+
+			lstate.joiner = make_unique<IEJoinUnion>(client, op, left_table, b1, right_table, b2);
+			return;
+		} else {
+			--next_pair;
+		}
+
+		// Outer joins
+		if (!left_outers && !right_outers) {
+			return;
+		}
+
+		// Spin wait for regular blocks to finish(!)
+		while (completed < pair_count) {
+			continue;
+		}
+
+		// Left outer blocks
+		const auto l = next_left++;
+		if (l < left_outers) {
+			lstate.left_block_index = l;
+			lstate.left_base = left_bases[l];
+
+			lstate.left_matches = left_table.found_match.get() + lstate.left_base;
+			lstate.outer_idx = 0;
+			lstate.outer_count = left_table.BlockSize(l);
+			return;
+		} else {
+			lstate.left_matches = nullptr;
+			--next_left;
+		}
+
+		// Right outer block
+		const auto r = next_right++;
+		if (r < right_outers) {
+			lstate.right_block_index = r;
+			lstate.right_base = right_bases[r];
+
+			lstate.right_matches = right_table.found_match.get() + lstate.right_base;
+			lstate.outer_idx = 0;
+			lstate.outer_count = right_table.BlockSize(r);
+			return;
+		} else {
+			lstate.right_matches = nullptr;
+			--next_right;
+		}
+	}
+
+	void PairCompleted(ClientContext &client, IEJoinGlobalState &gstate, IEJoinLocalSourceState &lstate) {
+		lstate.joiner.reset();
+		++completed;
+		GetNextPair(client, gstate, lstate);
+	}
+
+	const PhysicalIEJoin &op;
+
+	mutex lock;
+	bool initialized;
+
+	// Join queue state
+	std::atomic<size_t> next_pair;
+	std::atomic<size_t> completed;
+
+	// Block base row number
+	vector<idx_t> left_bases;
+	vector<idx_t> right_bases;
+
+	// Outer joins
+	idx_t left_outers;
+	std::atomic<idx_t> next_left;
+
+	idx_t right_outers;
+	std::atomic<idx_t> next_right;
+};
+
+unique_ptr<GlobalSourceState> PhysicalIEJoin::GetGlobalSourceState(ClientContext &context) const {
+	return make_unique<IEJoinGlobalSourceState>(*this);
+}
+
+unique_ptr<LocalSourceState> PhysicalIEJoin::GetLocalSourceState(ExecutionContext &context,
+                                                                 GlobalSourceState &gstate) const {
+	return make_unique<IEJoinLocalSourceState>(*this);
+}
+
+void PhysicalJoin::ConstructRightJoinResult(DataChunk &right, DataChunk &result, bool found_match[]) {
+	SelectionVector remaining_sel(STANDARD_VECTOR_SIZE);
+	idx_t remaining_count = 0;
```

Comment:
This function appears to be unused

Reply:
Yes, it was left over from a previous (slow!) right join implementation.  Should I remove it or move it next to the corresponding left join constructor?